src/main/java/core/basesyntax/ArrayList.java

```
@@ -1,48 +1,134 @@
+        T[] newArr = (T[]) new Object[capacity];
+        if (index > 0) {
+            System.arraycopy(classList, 0, newArr, 0, index);
+        }
+        if (index <= size()) {
+            System.arraycopy(classList, index, newArr, index + 1, size() - index);
+        }
+        newArr[index] = value;
+        classList = newArr;
+        nextFreeIndex++;
     }
 
+    @SuppressWarnings("unchecked")
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIfIndexExist(index);
+        return classList[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIfIndexExist(index);
+        classList[index] = value;
     }
 
+    @SuppressWarnings("unchecked")
     @Override
     public T remove(int index) {
-        return null;
+        checkIfIndexExist(index);
+        T[] newArr = (T[]) new Object[capacity];
+        if (index > 0) {
+            System.arraycopy(classList, 0, newArr, 0, index);
+        }
+        if (index < size()) {
+            System.arraycopy(classList, index + 1, newArr, index, size() - index - 1);
+        }
+        T removedObject = get(index);
+        classList = newArr;
+        nextFreeIndex--;
+        return removedObject;
+
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        boolean foundElement = false;
+        for (int i = 0; i < size(); i++) {
+            if (element == null && classList[i] == null) {
+                remove(i);
+                foundElement = true;
+                break;
+            }
+            if (classList[i] != null && classList[i].equals(element)) {
+                remove(i);
+                foundElement = true;
+                break;
+            }
+        }
+        if (!foundElement) {
+            throw new NoSuchElementException();
```

This violates the checklist item: "Make your exceptions informative. If you throw an Exception, you should always add an informative message so that it would be clear from the stack trace what exactly went wrong." The code throws a NoSuchElementException with no message in remove(T element): `throw new NoSuchElementException();`. Please include a clear message (for example: `throw new NoSuchElementException("Element not found: " + element);`) so the exception is informative and meets the checklist requirement. Reference: checklist text (Make your exceptions informative...)  and your ArrayList implementation where the exception is thrown .